src/ServiceControl.Monitoring/Http/Diagrams/DiagramApiModule.cs

```diff
@@ -81,22 +86,43 @@ public MonitoredEndpointsModule(IProvideBreakdownBy<EndpointInstanceId>[] metric
                     digest.Metrics.Add(metric.ReturnName, new MonitoredEndpointMetricDigest{Latest = values.Points.LastOrDefault(), Average = values.Average});
                 }
 
-                foreach (var metric in instanceMetric)
+                var metricDetails = new MonitoredEndpointMetricDetails();
+
+                foreach (var metric in instanceMetrics)
                 {
                     var store = metricByInstanceLookup[metric.StoreType];
-                    var intervals = store.GetIntervals(period, DateTime.UtcNow).ToLookup(k => k.Id.InstanceId);
+                    var intervals = store.GetIntervals(period, DateTime.UtcNow);
+                    var intervalsByInstanceId = intervals.ToLookup(k => k.Id.InstanceId);
+                    var intervalsByEndpoint = intervals.ToLookup(k => k.Id.EndpointName);
 
                     foreach (var instance in instances)
                     {
-                        var values = metric.Aggregate(intervals[instance.Id].ToList(), period);
+                        var values = metric.Aggregate(intervalsByInstanceId[instance.Id].ToList(), period);
 
                         instance.Metrics.Add(metric.ReturnName, values);
                     }
+
+                    if (detailedMetrics.Contains(metric.ReturnName))
+                    {
+                        var detailedValues = metric.Aggregate(intervalsByEndpoint[endpointName].ToList(), period);
```

Comment:
`metric.Aggregate(intervals[endpointName].ToList(), period);` is the exactly the same code that is executed in the first loop. This could and should be used only once.